runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactoryTest.java

```diff
@@ -459,27 +461,28 @@ public void loadBalancesBundles() throws Exception {
       verify(envFactory, Mockito.times(2)).createEnvironment(eq(environment), any());
 
       long tms = System.currentTimeMillis();
-      AtomicBoolean closed = new AtomicBoolean();
-      // close to free up environment for another bundle
-      TimerTask closeBundleTask =
-          new TimerTask() {
-            @Override
-            public void run() {
-              try {
-                b2.close();
-                closed.set(true);
-              } catch (Exception e) {
-                throw new RuntimeException(e);
-              }
-            }
-          };
-      new Timer().schedule(closeBundleTask, 100);
-
+      ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+      ScheduledFuture<Optional<Exception>> closingFuture =
+          executor.schedule(
+              () -> {
+                try {
+                  b2.close();
+                  return Optional.empty();
+                } catch (Exception e) {
+                  return Optional.of(e);
+                }
+              },
+              100,
+              TimeUnit.MILLISECONDS);
+
+      // This call should block until closingFuture has finished closing b2
       RemoteBundle b3 = sbf.getBundle(orf, srh, BundleProgressHandler.ignored());
 
-      // ensure we waited for close
+      // ensure the previous call waited for close
       Assert.assertThat(System.currentTimeMillis() - tms, greaterThanOrEqualTo(100L));
-      Assert.assertThat(closed.get(), is(true));
+      // This assertion includes a small delay to give the forked thread a chance to finish if it's
+      // been blocked
+      Assert.assertThat(closingFuture.get(1, TimeUnit.MILLISECONDS), equalTo(Optional.empty()));
```

Comment:
nit: Small delays lead to flakes, for example if the system is busy and doesn't get back to the forked thread. Does having a small timeout add any value here? Could you just block?

Reply:
Yeah agree a small delay isn't ideal.. my concern with blocking indefinitely is that it could return much later and the test would still pass which wouldn't be correct. It _should_ be done.

Maybe I'm just splitting hairs, the prior assertion on System time is pretty strong so we could just do without this one.

Reply:
That does seem like a good reason to want a timeout.

Here are some possible ways to tighten things down further:
1. Set `tms` in the thread just before calling `close()`, then check `getBundle` returns after that time. (This guards against the system is running slow and the thread scheduling after significantly more than 100ms.)
2. Make the delay of 100 ms random. (This with 1 guards against a `Thread.sleep(100L)` implementation of `getBundle`.)

Reply:
Ok I did similar to your (1), but rather than setting tms I just set an AtomicBoolean prior to calling close, and verify it is set right after getBundle returns.